ShaderTool/Command/Map.cs

```diff
@@ -531,41 +544,48 @@ class Map {
                 float actorXScale = actorData.localTransform[0];
                 float actorYScale = actorData.localTransform[5];
 
-                // List is always a multiple of 4
-                int actorTransformID = transformList.Count / 4;
-                StreamWrite(actorTransformID);
                 transformList.AddRange(new float[] { actorX, actorY, actorXScale, actorYScale });
 
-                // Instances
-                StreamWrite(actorData.instances.Length);
-                int instanceCount = 0;
                 foreach (Instance instance in actorData.instances) {
-                    StreamWrite(instanceCount); // used as ID here
-
                     float[] instanceMatrix = CalculateInstanceMatrix(actorData, instance); // { x, y, xScale, yScale }
-                    foreach (float value in instanceMatrix)
-                        StreamWrite(value);
 
-                    int instanceTransformID = transformList.Count / 4;
-                    StreamWrite(instanceTransformID);
+                    uint instanceTransformID = (uint)transformList.Count / 4;
+                    _TransformData data = new _TransformData();
+                    data.transform = instance.matrix;
+                    data.transformId = instanceTransformID;
+                    data.animationId = 0;
+                    instanceList.Add(data);
 
                     transformList.AddRange(instanceMatrix);
                 }
 
             }
 
-            StreamWrite(0xFFFFFFFF);
+            StreamWrite(instanceList.Count);
```

Comment:
you already wrote the instance count in line 524 though? why was the max int removed?

Reply:
To preallocate memory in the actor read function.